insights/config/specs.py

```diff
@@ -67,6 +67,8 @@
     "dirsrv"                    : SimpleFileSpec("etc/sysconfig/dirsrv"),
     "dirsrv_access"             : PatternSpec("var/log/dirsrv/.*/access"),
     "dirsrv_errors"             : PatternSpec("var/log/dirsrv/.*/errors"),
+    "disk_space"                : All([SimpleFileSpec("mongodb_disk_space"),
+                                    SimpleFileSpec("postgres_disk_space")]),
```

Comment:
Currently in 1.x, regardless of whether this is a PatternSpec or a SimpleFileSpec, only the first file found will be collected.  So you can split this into two specs or wait until the problem is corrected in 1.x.

Reply:
Aren't these SimpleFileSpecs looking for a file called 'mongodb_disk_space' in the root directory (of the client file system, or the sosreport)?  Shouldn't they be an actual command?  How do these files actually get generated?

Reply:
@bfahr - I have made a test for PatternSpec and SimpleFileSpec in `All` and `NoneGroup`, the result is:
1. No matter PatternSpec or SimpleFileSpec, if there are put in an `All` function, all the existed files will be collected by our client_1.x.
2. **However**, for All([SimpleFileSpec()..]), only the first file will be parsed by the framework, all others are just skipped.
3. For `All([PatternSpec()..]) or NoneGroup([PatternSpec()..])`, all the collected files will be parsed, but be careful, the parsed result will be **wrapped into a list**, so it's necessary for rule developer to iterate the list to get each file's result.

@Sheersha-jain, for the `NoneGroup` we talked on IRC yesterday, it's used to indicate the collector do not attempt to collect the files in it.
So, here you can 
1). use PatternSpec instead of SimpleFileSpec, and replace the `All` to `NoneGroup`.
2). continue using SimpleFileSpec, but split them into two `NoneGroup` specs.
I prefer to choose 1), since the two files are already collected by sosreport or foreman-debug and they are basically in the same format and can be parsed by one Parser by using one `spec`
How do you think it, @Bob?

@PaulWay - Just FYI, these two files can be found in sosreport/foreman-debug archive when corresponding files exist in the system.

```
[root@liuxc-rhel7-sat6 ~]# find -name "*_disk_space*"                                          
./sosreport-liuxc-rhel7-sat6-20171106112737/sos_commands/foreman/foreman-debug/postgres_disk_space                                                                                             
./sosreport-liuxc-rhel7-sat6-20171106112737/sos_commands/foreman/foreman-debug/mongodb_disk_space                                                                                              
./foreman-debug-HBR7p/postgres_disk_space      
./foreman-debug-HBR7p/mongodb_disk_space       
[root@liuxc-rhel7-sat6 ~]# cat ./foreman-debug-HBR7p/mongodb_disk_space
COMMAND> du -sh /var/lib/mongodb

5.1G    /var/lib/mongodb
[root@liuxc-rhel7-sat6 ~]# cat ./sosreport-liuxc-rhel7-sat6-20171106112737/sos_commands/foreman/foreman-debug/postgres_disk_space
COMMAND> du -sh /var/lib/pgsql

216M    /var/lib/pgsql
```
IMHO, I think we can add these two commands "du -sh /var/lib/mongodb(or pgsql)" to the `specs.py` when there are some public rules need them in the future.

Reply:
@xiangce Then the correct way AFAIK to do that is with the specs:
```python
    "mongodb_disk_space"                : First([SimpleFileSpec("mongodb_disk_space"), CommandSpec("du -sh /var/lib/mongodb")]),
    "postgres_disk_space"                : First([SimpleFileSpec("postgres_disk_space"), CommandSpec("du -sh /var/lib/pgsql")]),
```

And then have one base parser class - e.g. `DiskUsage` - and two parsers, say `MongoDBDiskUsage` and `PostgresDiskUsage` based off that base class.

A path that's specific to Red Hat's way of gathering customer data in a sosreport is useless for the rest of the world that doesn't use a sosreport.  It's useless for the customers we have that use Insights rather than the CLI.  So I would really strongly encourage @Sheersha-jain to consider all users of this parser, not only those based on reading a sosreport.

Reply:
@PaulWay - Yep, splitting them into two also makes sense to me. And in your example, it easily achieved to support the sosreport and insights-data.

**But**, you should put the CommandSpec in the first place, since the insights-client will only collect the first command/simple-file for `First`:

```
    "mongodb_disk_space"                : First([CommandSpec("du -sh /var/lib/mongodb"), SimpleFileSpec("mongodb_disk_space")]),
    "postgres_disk_space"               : First([CommandSpec("du -sh /var/lib/pgsql"), SimpleFileSpec("postgres_disk_space")]),

```

Reply:
@xiangce @PaulWay Even if an All group has a PatternSpec, only the first one that produces a result is evaluated by the engine in 1.x. [The code](https://github.com/RedHatInsights/insights-core/blob/1.x/insights/core/specs.py#L106) breaks after the first match. The [intent of the All group](https://github.com/RedHatInsights/insights-core/blob/1.x/insights/config/__init__.py#L309) is to cover cases where the same file(s) might be in different places in different versions of RHEL, so the collector is instructed to collect everything in an All group, but in reality the engine expects only one of the individual entries to be in a given archive.

Reply:
@csams So what's the difference between the 'First' and 'All' groups?
@xiangce Yes, that'd be better to put the command first :-)

Reply:
@PaulWay I think need to specify "/usr/bin" in command specs 
```
    "mongodb_disk_space"        : First([CommandSpec("/usr/bin du -sh /var/lib/mongodb"), SimpleFileSpec("mongodb_disk_space")]),
    "postgres_disk_space"       : First([CommandSpec("/usr/bin du -sh /var/lib/pgsql"), SimpleFileSpec("postgres_disk_space")]),
```

Reply:
@csams - Yes, my fault. The `All([PatternSpecs..]) only works in unit test :(. I tested it with the `insights-cli` tool, it indeed only parsed the first file.

Reply:
And I have the same question: what's the difference between `First` and `All`?  :((